In [ ]:
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Intro to Generating and Executing Python Code with Gemini 2.0

<table align="left">
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo"><br> Open in Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https:%2F%2Fraw.githubusercontent.com%2FGoogleCloudPlatform%2Fgenerative-ai%2Fmain%2Fgemini%2Fcode-execution%2Fintro_code_execution.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo"><br> Open in Colab Enterprise
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/GoogleCloudPlatform/generative-ai/main/gemini/code-execution/intro_code_execution.ipynb">
      <img src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo"><br> Open in Vertex AI Workbench
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery Studio logo"><br> Open in BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://upload.wikimedia.org/wikipedia/commons/9/91/Octicons-mark-github.svg" alt="GitHub logo"><br> View on GitHub
    </a>
  </td>
</table>

<div style="clear: both;"></div>

<b>Share to:</b>

<a href="https://www.linkedin.com/sharing/share-offsite/?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="LinkedIn logo">
</a>

<a href="https://bsky.app/intent/compose?text=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="Bluesky logo">
</a>

<a href="https://twitter.com/intent/tweet?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/53/X_logo_2023_original.svg" alt="X logo">
</a>

<a href="https://reddit.com/submit?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit logo">
</a>

<a href="https://www.facebook.com/sharer/sharer.php?u=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/51/Facebook_f_logo_%282019%29.svg" alt="Facebook logo">
</a>

| | |
|-|-|
| Author(s) |  [Kristopher Overholt](https://github.com/koverholt/) |

## Overview

This notebook introduces the code execution capabilities of the [Gemini 2.0 Flash model](https://cloud.google.com/vertex-ai/generative-ai/docs/gemini-v2), a new multimodal generative AI model from Google [DeepMind](https://deepmind.google/). Gemini 2.0 Flash offers improvements in speed, quality, and advanced reasoning capabilities including enhanced understanding, coding, and instruction following.

## Code Execution

A key feature of this model is [code execution](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution), which is the ability to generate and execute Python code directly within the API. If you want the API to generate and run Python code and return the results, you can use code execution as demonstrated in this notebook.

This code execution capability enables the model to generate code, execute and observe the results, correct the code if needed, and learn iteratively from the results until it produces a final output. This is particularly useful for applications that involve code-based reasoning such as solving mathematical equations or processing text.

## Objectives

In this tutorial, you will learn how to generate and execute code using the Gemini API in Vertex AI and the Google Gen AI SDK for Python with the Gemini 2.0 Flash model.

You will complete the following tasks:

- Generating and running sample Python code from text prompts
- Exploring data using code execution in multi-turn chats
- Using code execution in streaming sessions

## Getting started

### Install Google Gen AI SDK for Python


In [1]:
%pip install --upgrade --quiet google-genai

### Authenticate your notebook environment (Colab only)

If you're running this notebook on Google Colab, run the cell below to authenticate your environment.

In [2]:
import sys

if "google.colab" in sys.modules:
    from google.colab import auth

    auth.authenticate_user()

### Import libraries

In [3]:
import os

from IPython.display import Markdown, display
from google import genai
from google.genai.types import GenerateContentConfig, Tool

### Connect to a generative AI API service

Google Gen AI APIs and models including Gemini are available in the following two API services:

- [Google AI for Developers](https://ai.google.dev/gemini-api/docs): Experiment, prototype, and deploy small projects.
- [Vertex AI](https://cloud.google.com/vertex-ai/generative-ai/docs/overview): Build enterprise-ready projects on Google Cloud.
The Google Gen AI SDK provides a unified interface to these two API services.

This notebook shows how to use the Google Gen AI SDK with the Gemini API in Vertex AI.

### Set Google Cloud project information and create client

To get started using Vertex AI, you must have an existing Google Cloud project and [enable the Vertex AI API](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com).

Learn more about [setting up a project and a development environment](https://cloud.google.com/vertex-ai/docs/start/cloud-environment).

In [4]:
PROJECT_ID = "[your-project-id]"  # @param {type: "string", placeholder: "[your-project-id]", isTemplate: true}
if not PROJECT_ID or PROJECT_ID == "[your-project-id]":
    PROJECT_ID = str(os.environ.get("GOOGLE_CLOUD_PROJECT"))

LOCATION = os.environ.get("GOOGLE_CLOUD_REGION", "us-central1")

In [5]:
client = genai.Client(vertexai=True, project=PROJECT_ID, location=LOCATION)

### Improve code rendering in cell outputs

In [6]:
from IPython.display import HTML, Markdown


# Modify CSS to display the results more clearly in Colab
def set_css_in_cell_output(unused):
    display(
        HTML(
            """<style>
div.output_markdown pre {
  font-size: 16px;
  background: #d2bef4;
  margin: 16px;
  padding: 16px;
}

div.output_markdown p {
  font-size: 16px;
  background: #b4dcfe;
  margin: 16px;
  padding: 16px;
  color: #222222;
}
</style>"""
        )
    )


get_ipython().events.register("pre_run_cell", set_css_in_cell_output)

## Working with code execution in Gemini 2.0

### Load the Gemini model

The following code loads the Gemini 2.0 Flash model. You can learn about all Gemini models on Vertex AI by visiting the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models):

In [7]:
MODEL_ID = "gemini-2.0-flash-exp"  # @param {type: "string"}

### Define the code execution tool

The following code initializes the code execution tool by passing `code_execution` in a `Tool` definition.

Later we'll register this tool with the model that it can use to generate and run Python code:

In [8]:
code_execution_tool = Tool(code_execution={})

### Generate and execute code

The following code sends a prompt to the Gemini model, asking it to generate and execute Python code to calculate the sum of the first 50 prime numbers. The code execution tool is passed in so the model can generate and run the code:

In [9]:
PROMPT = """
What is the sum of the first 50 prime numbers?
Generate and run code for the calculation.
"""

response = client.models.generate_content(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

### View the generated code

The following code iterates through the response and displays any generated Python code by checking for `part.executable_code` in the response parts:

In [10]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```
{part.executable_code.code}
```
"""
            )
        )


```

def is_prime(n):
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

primes = []
num = 2
while len(primes) < 50:
    if is_prime(num):
        primes.append(num)
    num += 1

sum_of_primes = sum(primes)
print(f'{sum_of_primes=}')

```


### View the code execution results

The following code iterates through the response and displays the execution result and outcome by checking for `part.code_execution_result` in the response parts:

In [11]:
for part in response.candidates[0].content.parts:
    if part.code_execution_result:
        display(Markdown(part.code_execution_result.output))
        print("\nOutcome:", part.code_execution_result.outcome)

sum_of_primes=5117



Outcome: OUTCOME_OK


Great! Now you have the answer (5117) as well as the generated (and verified via execution!) Python code.

At this point in your application, you would save the output code, result, or outcome and display it to the end-user or use it downstream in your application.

### Code execution in a chat session

This section shows how to use code execution in an interactive chat with history using the Gemini API.

You can use `client.chats.create` to create a chat session and passes in the code execution tool, enabling the model to generate and run code:

In [12]:
chat = client.chats.create(
    model=MODEL_ID,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

You'll start the chat by asking the model to generate sample time series data with noise and then output a sample of 10 data points:

In [13]:
response = chat.send_message(
    """Generate code that creates sample time series
data of temperature vs. time in a test furnace. Add noise to the data. Output
a sample of 10 data points from the time series data."""
)

Now you can iterate through the response to display any generated Python code and execution results by checking for `part.executable_code` and `part.code_execution_result` in the response parts:

In [14]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(part.code_execution_result.output))
        print("\nOutcome:", part.code_execution_result.outcome)


```

import numpy as np

# 1. Define Time Range
time = np.linspace(0, 10, 100)  # 100 points from 0 to 10 seconds

# 2. Generate Base Temperature Data (linear increase)
base_temp = 20 + 5 * time  # Start at 20 degrees, increase by 5 degrees per second

# 3. Add Noise
noise = np.random.normal(0, 2, len(time))  # Gaussian noise with mean 0, std dev 2
noisy_temp = base_temp + noise

# 4. Output Sample
sample_indices = np.linspace(0, len(time) - 1, 10, dtype=int)
sample_time = time[sample_indices]
sample_temp = noisy_temp[sample_indices]

print("Sample Time Series Data (Time, Temperature):")
for t, temp in zip(sample_time, sample_temp):
    print(f"Time: {t:.2f} s, Temperature: {temp:.2f} °C")

```


Sample Time Series Data (Time, Temperature):
Time: 0.00 s, Temperature: 21.10 °C
Time: 1.11 s, Temperature: 27.38 °C
Time: 2.22 s, Temperature: 32.54 °C
Time: 3.33 s, Temperature: 35.23 °C
Time: 4.44 s, Temperature: 44.09 °C
Time: 5.56 s, Temperature: 49.99 °C
Time: 6.67 s, Temperature: 52.68 °C
Time: 7.78 s, Temperature: 59.13 °C
Time: 8.89 s, Temperature: 64.07 °C
Time: 10.00 s, Temperature: 66.55 °C



Outcome: OUTCOME_OK


Now you can ask the model to add a smoothed data series to the time series data:

In [15]:
response = chat.send_message(
    """Now add a data series that smooths the data using an appropriate method."""
)

And then display the generated Python code and execution results:

In [16]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(part.code_execution_result.output))
        print("\nOutcome:", part.code_execution_result.outcome)


```

import numpy as np

def moving_average(data, window_size):
    """Calculates the moving average of a 1D array."""
    if window_size > len(data):
        raise ValueError("Window size cannot be larger than the data length.")
    
    weights = np.repeat(1.0, window_size) / window_size
    return np.convolve(data, weights, 'valid')

# 1. Define Time Range
time = np.linspace(0, 10, 100)  # 100 points from 0 to 10 seconds

# 2. Generate Base Temperature Data (linear increase)
base_temp = 20 + 5 * time  # Start at 20 degrees, increase by 5 degrees per second

# 3. Add Noise
noise = np.random.normal(0, 2, len(time))  # Gaussian noise with mean 0, std dev 2
noisy_temp = base_temp + noise

# 4. Calculate Moving Average
window_size = 5
smoothed_temp = moving_average(noisy_temp, window_size)

# Adjust time array to match the length of smoothed data
smoothed_time = time[window_size - 1:]

# 5. Output Sample
sample_indices = np.linspace(0, len(smoothed_time) - 1, 10, dtype=int)
sample_time = smoothed_time[sample_indices]
sample_noisy_temp = noisy_temp[window_size - 1:][sample_indices]
sample_smoothed_temp = smoothed_temp[sample_indices]


print("Sample Time Series Data (Time, Noisy Temp, Smoothed Temp):")
for t, noisy_temp, smoothed_temp in zip(sample_time, sample_noisy_temp, sample_smoothed_temp):
    print(f"Time: {t:.2f} s, Noisy Temp: {noisy_temp:.2f} °C, Smoothed Temp: {smoothed_temp:.2f} °C")

```


Sample Time Series Data (Time, Noisy Temp, Smoothed Temp):
Time: 0.40 s, Noisy Temp: 24.24 °C, Smoothed Temp: 20.87 °C
Time: 1.41 s, Noisy Temp: 25.06 °C, Smoothed Temp: 25.35 °C
Time: 2.53 s, Noisy Temp: 35.70 °C, Smoothed Temp: 31.72 °C
Time: 3.54 s, Noisy Temp: 37.72 °C, Smoothed Temp: 37.53 °C
Time: 4.65 s, Noisy Temp: 42.47 °C, Smoothed Temp: 41.59 °C
Time: 5.66 s, Noisy Temp: 46.74 °C, Smoothed Temp: 47.52 °C
Time: 6.77 s, Noisy Temp: 51.56 °C, Smoothed Temp: 52.91 °C
Time: 7.78 s, Noisy Temp: 59.30 °C, Smoothed Temp: 57.67 °C
Time: 8.89 s, Noisy Temp: 63.41 °C, Smoothed Temp: 62.32 °C
Time: 10.00 s, Noisy Temp: 69.06 °C, Smoothed Temp: 69.32 °C



Outcome: OUTCOME_OK


Finally, you can ask the model to generate descriptive statistics for the time series data:

In [17]:
response = chat.send_message(
    """Now generate and output descriptive statistics on the time series data."""
)

And then display the generated Python code and execution results:

In [18]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(part.code_execution_result.output))
        print("\nOutcome:", part.code_execution_result.outcome)


```

import numpy as np

def moving_average(data, window_size):
    """Calculates the moving average of a 1D array."""
    if window_size > len(data):
        raise ValueError("Window size cannot be larger than the data length.")
    
    weights = np.repeat(1.0, window_size) / window_size
    return np.convolve(data, weights, 'valid')

# 1. Define Time Range
time = np.linspace(0, 10, 100)  # 100 points from 0 to 10 seconds

# 2. Generate Base Temperature Data (linear increase)
base_temp = 20 + 5 * time  # Start at 20 degrees, increase by 5 degrees per second

# 3. Add Noise
noise = np.random.normal(0, 2, len(time))  # Gaussian noise with mean 0, std dev 2
noisy_temp = base_temp + noise

# 4. Calculate Moving Average
window_size = 5
smoothed_temp = moving_average(noisy_temp, window_size)

# Adjust time array to match the length of smoothed data
smoothed_time = time[window_size - 1:]

# 5. Calculate Descriptive Statistics
noisy_mean = np.mean(noisy_temp)
noisy_std = np.std(noisy_temp)
noisy_min = np.min(noisy_temp)
noisy_max = np.max(noisy_temp)

smoothed_mean = np.mean(smoothed_temp)
smoothed_std = np.std(smoothed_temp)
smoothed_min = np.min(smoothed_temp)
smoothed_max = np.max(smoothed_temp)


# 6. Output Statistics
print("Descriptive Statistics:")
print("--------------------------------------------------")
print("Noisy Temperature Data:")
print(f"  Mean: {noisy_mean:.2f} °C")
print(f"  Standard Deviation: {noisy_std:.2f} °C")
print(f"  Minimum: {noisy_min:.2f} °C")
print(f"  Maximum: {noisy_max:.2f} °C")
print("--------------------------------------------------")
print("Smoothed Temperature Data:")
print(f"  Mean: {smoothed_mean:.2f} °C")
print(f"  Standard Deviation: {smoothed_std:.2f} °C")
print(f"  Minimum: {smoothed_min:.2f} °C")
print(f"  Maximum: {smoothed_max:.2f} °C")
print("--------------------------------------------------")

```


Descriptive Statistics:
--------------------------------------------------
Noisy Temperature Data:
  Mean: 44.80 °C
  Standard Deviation: 14.48 °C
  Minimum: 17.34 °C
  Maximum: 70.30 °C
--------------------------------------------------
Smoothed Temperature Data:
  Mean: 44.84 °C
  Standard Deviation: 13.84 °C
  Minimum: 21.89 °C
  Maximum: 67.61 °C
--------------------------------------------------



Outcome: OUTCOME_OK


This chat example demonstrates how you can use the Gemini API with code execution as a powerful tool for exploratory data analysis and more. Go forth and adapt this approach to your own projects and use cases!

### Code execution in a streaming session

You can also use the code execution functionality with streaming output from the Gemini API.

The following code demonstrates how the Gemini API can generate and execute code while streaming the results:

In [19]:
PROMPT = """
Generate a list of 20 random names, then create a new list with just the names
containing the letter 'a', then output the number of names that contain 'a' and
finally show me that new list.
"""

for chunk in client.models.generate_content_stream(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
):
    for part in chunk.candidates[0].content.parts:
        if part.text:
            display(Markdown("#### Natural language stream"))
            display(Markdown(part.text))
            display(Markdown("---"))
        if part.executable_code:
            display(Markdown("#### Code stream"))
            display(
                Markdown(
                    f"""
```
{part.executable_code.code}
```
"""
                )
            )
            display(Markdown("---"))
        if part.code_execution_result:
            display(Markdown("#### Code result"))
            display(
                Markdown(
                    f"""
```
{part.code_execution_result.output}
```
"""
                )
            )
            display(Markdown("---"))

#### Natural language stream

Okay

---

#### Natural language stream

, I can do that. Here's how I'll approach this:

---

#### Natural language stream



1.  **Generate 20 random names:** I'll use

---

#### Natural language stream

 Python's `random` module to generate a list of 20 random names. For simplicity, I'll use a combination of common first names.

---

#### Natural language stream


2.  **Filter for names with 'a':** I'll iterate through the list and create a new list containing only the names that include the

---

#### Natural language stream

 letter 'a' (case-insensitive).
3.  **Count and output:** I'll count the number of names in the filtered list and output that count, along with the filtered list itself.

Here's the code

---

#### Natural language stream

:



---

#### Code stream


```

import random

def generate_random_names(num_names):
    first_names = ["Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Grace", "Henry", "Ivy", "Jack", "Kate", "Liam", "Mia", "Noah", "Olivia", "Peter", "Quinn", "Ryan", "Sophia", "Tom"]
    return random.choices(first_names, k=num_names)

def filter_names_with_a(names):
    return [name for name in names if 'a' in name.lower()]

# Generate 20 random names
random_names = generate_random_names(20)

# Filter names containing 'a'
names_with_a = filter_names_with_a(random_names)

# Count the names with 'a'
count_of_names_with_a = len(names_with_a)

# Output the results
print(f'{random_names=}')
print(f'{count_of_names_with_a=}')
print(f'{names_with_a=}')

```


---

#### Code result


```
random_names=['Noah', 'Bob', 'Tom', 'Quinn', 'Jack', 'Ryan', 'Henry', 'Eve', 'Kate', 'Liam', 'Ivy', 'Ivy', 'Eve', 'Henry', 'Liam', 'Jack', 'Bob', 'Frank', 'Grace', 'Kate']
count_of_names_with_a=10
names_with_a=['Noah', 'Jack', 'Ryan', 'Kate', 'Liam', 'Liam', 'Jack', 'Frank', 'Grace', 'Kate']

```


---

This streaming example demonstrated how the Gemini API can generate, execute code, and provide results within a streaming session.

## Summary

Refer to the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution) for more details about code execution, and in particular, the [recommendations](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution#code-execution-vs-function-calling) regarding differences between code execution and [function calling](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling).

### Next steps

- See the [Google Gen AI SDK reference docs](https://googleapis.github.io/python-genai/)
- Explore other notebooks in the [Google Cloud Generative AI GitHub repository](https://github.com/GoogleCloudPlatform/generative-ai)
- Explore AI models in [Model Garden](https://cloud.google.com/vertex-ai/generative-ai/docs/model-garden/explore-models)